### ISSUE TITLE
DevEnv takes a long time to initialize

### ISSUE THREAD

**Issue by MH-ZShearer** [actions/runner-images#5301]
### Description

In order to build a Visual Studio Installer Project (`.vdproj`), Visual Studio must be ran via `DevEnv.exe`. When building the installer, it takes anywhere from 2-10. This appears to be regression from #128. I'm not sure when the regression initially occurred as this is the first time I needed DevEnv.

In the reproduction steps below, I use `seanmiddleditch/gha-setup-vsdevenv@v4`. This appears to be required to ensure the installer is built due to the behavior of DevEnv.exe. It seems likely that you may be able to start and wait for the DevEnv process to finish, but I doubt that will result in a faster build. This action does add a lot of clutter to the environment variables as well.

Unfortunately, all our repositories are private and can't be shared publicly. For troubleshooting, I attempted to test our Action against windows-2019 based on #128. I experienced the same issue, as mentioned in Actual Behavior.

### Virtual environments affected

- [ ] Ubuntu 18.04
- [ ] Ubuntu 20.04
- [ ] macOS 10.15
- [ ] macOS 11
- [ ] Windows Server 2016
- [X] Windows Server 2019
- [X] Windows Server 2022

### Image version and build link

windows-2019 -> 20220316.1
windows-2022 -> 20220316.1

### Is it regression?

Maybe, see description

### Expected behavior

I expected about 20-30 seconds, but I'd be happy with less than 120 seconds.

### Actual behavior

It takes at least 2 minutes to **start** DevEnv, but typically takes 5-10 minutes. The last two runs clocked in at 8:40 and 8:28 on windows-2022 and windows-2019, respectively.

### Repro steps

``` yaml
name: DevEnv Test
on: workflow_dispatch
jobs:
  build:
    runs-on: windows-latest
    steps:
    - name: Setup MSBuild.exe
      uses: microsoft/setup-msbuild@v1.0.2

    - name: Setup VS Dev Environment
      uses: seanmiddleditch/gha-setup-vsdevenv@v4

    - name: Restore NuGet Packages
      run: msbuild /t:Restore

    - name: Build the application
      run: msbuild /t:Build /p:Configuration=Release

    # Required to Build VS Installer Projects
    - name: Disable Out of Process Building
      run: .\DisableOutOfProcBuild.exe
      working-directory: C:\Program Files\Microsoft Visual Studio\2022\Enterprise\Common7\IDE\CommonExtensions\Microsoft\VSI\DisableOutOfProcBuild

    # Can take 10+ minutes to complete
    - name: Build the Installer
      run: DevEnv Installer/Installer.vdproj /build Release
```

**Comment by al-cheb** [comment]
Hello @MH-ZShearer ,
We will take a look at it.

**Comment by flameshikari** [comment]
Hello there, I have the same problem. Any progress?

**Comment by al-cheb** [comment]
> Hello there, I have the same problem. Any progress?

Currently, no updates.

**Comment by domenichelfenstein** [comment]
I'm also having the same problem. Initialization takes 10min on `windows-2022`.
Doing the exact same thing as @MH-ZShearer (building an installer)

**Comment by delta42** [comment]
In the same boat

**Comment by f4l0** [comment]
Same on my side...

**Comment by ReMinoer** [comment]
Same here. I'm building an installer with devenv and it can easily take 10+ minutes.

**Comment by ilia-shipitsin** [comment]
new image has been delivered.
I'm closing this issue, feel free to reopen if is still a concern

**Comment by martinmueller4voice** [comment]
Thanks for taking your time, but I’m afraid it still takes very long before the project starts building.

I’ve just rebuilt my package.
The runner had version “2.305.0” and was using a Windows Server 2022 (10.0.20348).
Image was “windows-2022” version “20230612.1”

Fri, 23 Jun 2023 11:43:47 GMT  Run devenv.com $env:Solution_Name /rebuild "$env:Configuration|x86" /project $env:Installer_Project /log
Fri, 23 Jun 2023 11:43:47 GMT
Fri, 23 Jun 2023 11:50:28 GMT
Fri, 23 Jun 2023 11:50:28 GMT  Microsoft Visual Studio 2022 Version 17.6.2.
Fri, 23 Jun 2023 11:50:28 GMT  Copyright (C) Microsoft Corp. All rights reserved.
Fri, 23 Jun 2023 11:50:28 GMT
Fri, 23 Jun 2023 11:50:28 GMT  The license for Visual Studio expires in 19 days.
Fri, 23 Jun 2023 11:50:28 GMT  Rebuild started...

As you can see, it again takes almost 7 minutes before VS creates its first output.

When I built the project 2 weeks ago on a runner version “20230606.1”, this was the output:
Mon, 12 Jun 2023 10:53:17 GMT           Run devenv.com $env:Solution_Name /rebuild "$env:Configuration|x86" /project $env:Installer_Project /log
Mon, 12 Jun 2023 11:00:01 GMT
Mon, 12 Jun 2023 11:00:01 GMT           Microsoft Visual Studio 2022 Version 17.6.2.

7 minutes.


Am I too impatient and have to wait for an image version newer than 20230612.1 (unfortunately, you didn’t mention what the “new” image version was) or didn’t the fix work as hoped for?

Best regards
Martin Müller


Von: ilia-shipitsin ***@***.***>
Gesendet: Freitag, 23. Juni 2023 13:32
An: actions/runner-images ***@***.***>
Cc: Martin Müller ***@***.***>; Comment ***@***.***>
Betreff: Re: [actions/runner-images] DevEnv takes a long time to initialize (Issue #5301)


new image has been delivered.
I'm closing this issue, feel free to reopen if is still a concern

**Comment by ilia-shipitsin** [comment]
it was a false alarm from my side.
I'll check next week

**Comment by martinmueller4voice** [comment]
Ok, no problem.
Got a bit too excited :)
I'm looking forward to your next update.

**Comment by mikhailkoliada** [comment]
Should be deployed now

**Comment by martinmueller4voice** [comment]
Great, thank you so much!!
Those are the run times with the last runner version:

Mon, 26 Jun 2023 11:37:21 GMT	Runner Image
Mon, 26 Jun 2023 11:37:21 GMT	  Image: windows-2022
Mon, 26 Jun 2023 11:37:21 GMT	  Version: 20230620.1.0

Mon, 26 Jun 2023 11:39:56 GMT	Run devenv.com $env:Solution_Name /rebuild "$env:Configuration|x86" /project $env:Installer_Project /log
Mon, 26 Jun 2023 11:39:56 GMT	  
Mon, 26 Jun 2023 11:41:16 GMT	Microsoft Visual Studio 2022 Version 17.6.4.
Mon, 26 Jun 2023 11:41:16 GMT	Copyright (C) Microsoft Corp. All rights reserved.
Mon, 26 Jun 2023 11:41:16 GMT	The license for Visual Studio expires in 24 days.
Mon, 26 Jun 2023 11:41:16 GMT	Rebuild started...

1:20 minutes is a lot better than the ~7 minutes it took before.

Thanks again!

**Comment by Cyberboss** [comment]
I just had a build take 10mins to start up: 
![image](https://github.com/actions/runner-images/assets/8171642/57d3b56e-90b6-4d95-b597-9a84062247dc)

https://github.com/tgstation/tgstation-server/actions/runs/5382159227/jobs/9767199703

**Comment by MH-ZShearer** [comment]
I've noticed no perceivable difference than before. It still takes several minutes to even start devenv. The last 3 runs took 6:23, 6:28, and 7:57 minutes each just to start devenv with the entire build taking roughly 3 minutes longer each.

@martinmueller4voice since it appears to be working for you, it may be possible that we're setting up our devenv a bit differently, if that can even make a difference. Does your workflow differ from my original post? Are you able to post at least the important parts of your workflow for comparison?

**Comment by martinmueller4voice** [comment]
Hmmm - I don't know what's going on...
Those are the build durations for the last 3 builds:
![image](https://github.com/actions/runner-images/assets/73161814/fde116b0-94d5-4ea8-b0e2-5a14a874a583)

The 6 1/2 minute run was the one that made me cheer.
But the build from last week took >10 minutes again, while everything I had changed was the tag.

The fast run was with runner Version: 20230620.1.0
The latest, slow run was with runner  Version: 20230625.1.0

So from my point of view, something changed between 20230620 and 20230625 that reverted the speedup again.

**Comment by ilia-shipitsin** [comment]
@martinmueller4voice , I've added "vdproj" warmup. Warmup itself is simple vdproj.
can you share your details ? some project that is slow

**Comment by martinmueller4voice** [comment]
My projecct mainly contains those steps to build the setup project:

\# Install the .NET 7 workload
    - name: Install .NET
      uses: actions/setup-dotnet@v3
      with:
        dotnet-version: 7.0.x

\# Install GitVersion
    - name: Install GitVersion
      uses: gittools/actions/gitversion/setup@v0.9.7
      with:
        versionSpec: '5.x'

\# Setup NuGet
    - name: Setup Nuget
      uses: NuGet/setup-nuget@v1.0.2

\# Add  MSBuild to the PATH: https://github.com/microsoft/setup-msbuild
    - name: Setup MSBuild.exe
      uses: microsoft/setup-msbuild@v1.0.2

\# Execute all unit tests in the solution
    - name: Execute unit tests
      run: dotnet test -c Debug # Or else it looks as if Release is selected and there the .vdproj cannot be loaded

\# Setup devenv environment 
    - name: Setup VS Dev Environment
      uses: seanmiddleditch/gha-setup-vsdevenv@v4

\# Copy asp.net 7 x86 prerequisite Bootstrapper package
    - name: Copy ASP.NET 7 x86 Bootstrapper package
      run: dotnet run --project $env:SetupHelper

\# Build installer
    - name: Build installer
      run: devenv.com $env:Solution_Name /rebuild "$env:Configuration|x86" /project $env:Installer_Project /log


As you can see, the last step is the first one where devenv is actually started.

With the exact same steps, the build using 20230620.1 took 1:19 min to start devenv.com:
Mon, 26 Jun 2023 11:39:57 GMT Run devenv.com $env:Solution_Name /rebuild "$env:Configuration|x86" /project $env:Installer_Project /log
Mon, 26 Jun 2023 11:41:16 GMT 
Mon, 26 Jun 2023 11:41:16 GMT Microsoft Visual Studio 2022 Version 17.6.4.

The build using the 20230625.1 runner took 5:13 min again to start devenv.com:
Wed, 05 Jul 2023 12:36:31 GMT Run devenv.com $env:Solution_Name /rebuild "$env:Configuration|x86" /project $env:Installer_Project /log
Wed, 05 Jul 2023 12:41:44 GMT 
Wed, 05 Jul 2023 12:41:44 GMT Microsoft Visual Studio 2022 Version 17.6.4.

**Comment by woodne** [comment]
Does this change based on the size of the runner we use or something? I am using larger 16 core runners for a build and when I just run `devenv.exe /updateconfiguration` it can take anywhere from like 1-2 minutes, up to as long as almost 20 mintes: 
<img width="1053" alt="image" src="https://github.com/actions/runner-images/assets/303276/29eaddbb-cbd7-41f3-8f25-6a1f5067eeb1">
Is this expected behavior if the runner image already had this command run?

Runner information:
```
Current runner version: '2.306.0'
Runner name: 'win-server2022-latest-16core_155c5158a61a'
Runner group name: 'win-server2022-latest'
Machine name: 'runner'
Operating System
  Microsoft Windows Server 2022
  10.0.20348
  Datacenter
Runner Image
  Image: windows-2022
  Version: 20230716.1.0
  Included Software: https://github.com/actions/runner-images/blob/win22/20230716.1/images/win/Windows2022-Readme.md
  Image Release: https://github.com/actions/runner-images/releases/tag/win22%2F20230716.1
```

**Comment by mikhailkoliada** [comment]
Images are the same, @maxim-lobanov any ideas?

**Comment by floydpreyes** [comment]
Can I ask if this is closed because it's a known issue?

**Comment by martinmueller4voice** [comment]
Since we already had short startup times at some point, I feel that getting this to work reliably is just a small step away, even though I cannot say for sure what @maxim-lobanov did exactly with those images.
I fear that, once it's a known issue, it will stay a known issue indefinitely.
Any other opinions?

**Comment by EduPonz** [comment]
I still see very long times (~5 min) to simple start a dev shell in github hosted runners. Is there any update on this?

**Comment by martinmueller4voice** [comment]
Because there doesn't seem to be any progress, I thought I'd suggest using a local (self-hosted) runner.
Recently, I had to switch to a self-hosted runner because of new code signing requirements where you have to have a hardware module physically connected to the build machine.
Since this local runner doesn't have to be provisioned every time with a partially initialized VS, this can be a huge time saver.
And the best part: Self-hosted runners are not billed by minute.

**Comment by EduPonz** [comment]
Hi @martinmueller4voice, thanks for your reply! Unfortunately, one of the reasons to migrate to github actions was to avoid maintaining servers for CI, so we'd like to avoid self-hosted runners if we can.

**Comment by benediktadams** [comment]
Same problem here another 3 months later, builds with devenv.exe take at least 5-10minutes to start up on a Github runner....

**Comment by DavidWarry** [comment]
End of Feb 2025, and this is still a problem for me as well. Building a trivial installer takes ten minutes, though the output of the task reports that it took just over 2 mins to actually build the installer.

**Comment by benediktadams** [comment]
Yeah unfortunately it seems this has been abandoned. I have since looked through various 3rd party runner services, to see if any had this issue resolved, but they all just use the upstream repo. 

I am using https://www.warpbuild.com for most of my runners and I am in contact with support there, and they're actually aware of the issue and trying to solve it, but so far to no avail. 

We also investigated the issue internally and tried building our own images from the upstream repos:

It seems like the Visual Studio install goes fine, but devenv is still slow the first time it's called.
To mitigiate this, it seems like they try to call it once using the script in post-gen: https://github.com/actions/runner-images/blob/main/images/windows/assets/post-gen/VSConfiguration.ps1 

However, this needs to be run on each server after boot, it cannot be warmed up while building the image.
And we're not actually convinced it's being run on GH Actions runners, since then it would:
- Delay the boot of the runner
- Actually work when we run our actions which it doesn't

I cannot find any reference to these scripts, other than a mention: 
https://github.com/actions/runner-images/blob/ae16afb6f2273242cd4efd97c46d317e4199981e/docs/create-image-and-azure-resources.md?plain=1#L269-L283

So I assume this is done in some proprietary MS infra setup

- The warmup doesn't actually work basically
- They made a patch to the problem, but it's not having any effect
- The script is even linking to this GitHub issue

So yeah, this is frustrating, that this issue is obviously being ignored. I've for now come to terms running my windows runners for an extra 5-10minutes on every Windows build, but it sure sucks being billed for it.
